File: brady/diameter_of_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    // O(n) time, O(n) space
    public int diameterOfBinaryTree(TreeNode root) {
        
        if (root == null) return 0;

        int[] diameter = new int[1];
        findDiameter(root, diameter);

        return diameter[0];
    }

    // Returns the height of the tree rooted at "root", and updates the diameter object if necessary
    public int findDiameter(TreeNode root, int[] diameter) {
        if (root == null) return 0;

        int leftDepth = findDiameter(root.left, diameter);
        int rightDepth = findDiameter(root.right, diameter);

        diameter[0] = Math.max(diameter[0], leftDepth + rightDepth);

        return Math.max(leftDepth, rightDepth) + 1;
    }
}

File: zain/diameter_of_binary_tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        diameter = 0

        def dfs(root):
            nonlocal diameter
            if not root:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)
            diameter = max(left + right, diameter)
            return 1 + max(left, right)

        dfs(root)
        return diameter
